### PR TITLE
Update goose to find endpoints by requesting version information befo…

### PR DESCRIPTION
**Pull request by hmlanigan**
…re calls

This is a step on the way to API updates to include neutron replacements for deprecated nova API calls.  There is a corresponding commit required by juju.

The POC was from: https://github.com/axw/goose/commit/a7c4a4ef9d936eb113e6ab1c0d2af4612467cedb

Rackspace and OpenStack get the same results with swift testservice (local and live).  Nova testservice has a bug so the live tests won't run there.  Identity (live) testservice tests have a permission problem I haven't been able to work out yet, that doesn't occur in real life.  juju deploy with this code and the corresponding juju PR works for both Rackspace and OpenStack.

### CODE REVIEW

File: client/client.go [hunk 1]
```diff
@@ -163,44 +167,213 @@ func makeURL(base string, parts []string) string {
...
-	url, err := c.MakeServiceURL(svcType, []string{apiCall})
+	url, err := c.MakeServiceURL(svcType, apiVersion, []string{apiCall})
 	if err != nil {
 		return
 	}
 	return c.sendRequest(method, url, c.Token(), requestData)
 }
 
-func (c *authenticatingClient) MakeServiceURL(serviceType string, parts []string) (string, error) {
+func (c *authenticatingClient) MakeServiceURL(serviceType, apiVersion string, parts []string) (string, error) {
```
Comment (axw) [inline client/client.go:208]: can you please add caching of the version-specific URLs? as it is, this will query the versions every single API request

**Review by axw — CHANGES_REQUESTED** [review]

File: client/client.go (hunk 1 above)
Comment (hmlanigan) [reply client/client.go:208]: added to client.go: getAPIVersions()

File: client/client.go [hunk 2]
```diff
@@ -163,44 +170,231 @@ func makeURL(base string, parts []string) string {
...
 	if !ok {
 		return "", errors.New("no endpoints known for service type: " + serviceType)
 	}
-	return makeURL(url, parts), nil
+	if serviceType == "object-store" {
+		// object-store has no versioning capability.
+		return makeURL(serviceURL, parts), nil
+	}
+	if apiVersion == "" {
+		// some tests don't include a version
```
Comment (wallyworld) [inline client/client.go]: Let's not add special code to account for poorly constructed tests. If the apiVersion is required, any tests should be updated accordingly.

File: client/client.go [hunk 3]
```diff
@@ -163,44 +170,231 @@ func makeURL(base string, parts []string) string {
...
+		return makeURL(serviceURL, parts), nil
+	}
+	if strings.Contains(serviceURL, c.serviceURLs["object-store"]) {
```
Comment (wallyworld) [inline client/client.go]: Isn't this handling of object-store service type done up above?

File: client/client.go [hunk 4]
```diff
@@ -163,44 +170,231 @@ func makeURL(base string, parts []string) string {
...
+		// a container endpoint won't have a version
+		return makeURL(serviceURL, parts), nil
+	}
+	requestedVersion, err := parseVersion(apiVersion)
```
Comment (wallyworld) [inline client/client.go:216]: can we move this down next to getVersionURL where it is used

File: client/client.go [hunk 5]
```diff
@@ -163,44 +170,231 @@ func makeURL(base string, parts []string) string {
...
+		return "", err
+	}
+	serviceURL, err = getAPIVersionURL(rootURL, serviceURLSuffix, versions, requestedVersion)
+	if err != nil {
+		return "", err
+	}
+	return makeURL(serviceURL, parts), nil
+}
+
+func getAPIVersionURL(rootURL *url.URL, serviceURLSuffix string, versions []apiVersionInfo, requested apiVersion) (string, error) {
```
Comment (wallyworld) [inline client/client.go]: i think there's enough functionality and parameters here to warrant a comment

File: client/client.go [hunk 6]
```diff
@@ -163,44 +170,231 @@ func makeURL(base string, parts []string) string {
...
+	if match == "" {
+		return "", fmt.Errorf("could not find matching URL")
+	}
+	versionURL := *rootURL
+	versionURL.Path = path.Join(versionURL.Path, match, serviceURLSuffix)
+	return versionURL.String(), nil
+}
+
+type apiVersion struct {
+	major int
```
Comment (wallyworld) [inline client/client.go]: if we are going to unmarshal from json, our convention is to ensure there are json tags defined

File: client/client.go [hunk 7]
```diff
@@ -163,44 +170,231 @@ func makeURL(base string, parts []string) string {
...
+	minor int
+}
+
+func (v *apiVersion) UnmarshalJSON(b []byte) error {
```
Comment (wallyworld) [inline client/client.go]: All this api version stuff probably is best placed in a separate file like apiversion.go

File: client/client.go [hunk 8]
```diff
@@ -163,44 +170,231 @@ func makeURL(base string, parts []string) string {
...
+		if _, err := parseVersion(pathParts[0]); err == nil {
+			pathParts = pathParts[1:]
+		}
+		url.Path = "/"
+	}
+
+	baseURL := url.String()
+
+	// make sure we haven't already received the version info
+	apiURLVersion, ok := c.apiURLVersions[baseURL]
```
Comment (wallyworld) [inline client/client.go]: This isn't go routine safe and should be protected by a mutex

File: client/client.go [hunk 9]
```diff
@@ -163,44 +170,231 @@ func makeURL(base string, parts []string) string {
...
+	if ok {
```
Comment (wallyworld) [inline client/client.go]: this can be inlined with the apiURLVersions look up

File: client/client.go [hunk 10]
```diff
@@ -163,44 +170,231 @@ func makeURL(base string, parts []string) string {
...
+		return apiURLVersion.rootURL, apiURLVersion.serviceURLSuffix, apiURLVersion.versions, nil
```
Comment (wallyworld) [inline client/client.go]: instead of returning (rootURL, serviceURLSuffix, versions) why not just return the apiURLVersion struct itself

File: client/client.go [hunk 11]
```diff
@@ -163,44 +170,231 @@ func makeURL(base string, parts []string) string {
...
+	}
+	requestData := &goosehttp.RequestData{
+		RespValue: &raw,
+		ExpectedStatus: []int{
+			http.StatusOK,
+			http.StatusMultipleChoices,
+		},
+	}
+	if err := c.sendRequest("GET", baseURL, c.Token(), requestData); err != nil {
+		return nil, "", nil, err
```
Comment (wallyworld) [inline client/client.go]: yeah, we should just return *apiURLVersion
then this would be return nil, err

File: client/client.go [hunk 12]
```diff
@@ -163,44 +170,231 @@ func makeURL(base string, parts []string) string {
...
+	}
+
+	// Some services respond with {"versions":[...]}, and
+	// some respond with {"versions":{"values":[...]}}.
+	var object interface{}
+	var versions []apiVersionInfo
+	if err := json.Unmarshal(raw.Versions, &object); err != nil {
```
Comment (wallyworld) [inline client/client.go]: all this ummarshal stuff should be in a separate helper function, possible in that separate file mentioned elsewhere

File: testservices/cmd/main.go [hunk 13]
```diff
@@ -67,5 +67,6 @@ func main() {
 	for _, u := range users.users {
 		p.AddUser(u.user, u.secret, "tenant")
 	}
+	fmt.Printf("main(): %s\n", serveAddr)
```
Comment (wallyworld) [inline testservices/cmd/main.go]: delete me? or log at debug level?

File: testservices/novaservice/service.go [hunk 14]
```diff
@@ -143,6 +150,53 @@ func New(hostURL, versionPath, tenantId, region string, identityService, fallbac
 	return novaService
 }
 
+func NewNovaOnly(cred *identity.Credentials, auth identity.AuthMode) (*Nova, []string) {
```
Comment (wallyworld) [inline testservices/novaservice/service.go]: nitpic - all exported methods should have a comment, here and a few other places

File: testservices/novaservice/service_http.go [hunk 15]
```diff
@@ -167,8 +167,9 @@ The resource could not be found.
 	}
 	errNoVersion = &errorResponse{
 		http.StatusOK,
-		`{"versions": [{"status": "CURRENT", "updated": "2011-01-21` +
-			`T11:33:21Z", "id": "v2.0", "links": [{"href": "$ENDPOINT$", "rel": "self"}]}]}`,
+		`{"versions": [` +
+			`{"id": "v2.0", "links": [{"href": "v2", "rel": "self"}], "status": "SUPPORTED", "updated": "2011-01-21T11:33:21Z"}]}`,
+		//`{"id": "v2.1", "links": [{"href": "v2.1", "rel": "self"}], "status": "CURRENT", "updated": "2013-07-23T11:33:21Z"}]}`,
```
Comment (wallyworld) [inline testservices/novaservice/service_http.go]: delete me?

**Review by wallyworld — CHANGES_REQUESTED** [review]
There needs to be unit test coverage of the getAPIVersions and getAPIVersionURL methods. There's a lot of logic / corner cases there, eg
// Some services respond with {"versions":[...]},
// some respond with {"versions":{"values":[...]}}.

Whether those methods are tested explicitly, or via curated calls to MakeServiceURL doesn't really matter, although the latter may be preferred.

File: client/client.go [hunk 16]
```diff
@@ -163,44 +170,231 @@ func makeURL(base string, parts []string) string {
 	return base + strings.Join(parts, "/")
 }
 
-func (c *client) MakeServiceURL(serviceType string, parts []string) (string, error) {
+func (c *client) MakeServiceURL(serviceType, apiVersion string, parts []string) (string, error) {
 	return makeURL(c.baseURL, parts), nil
```
Comment (wallyworld) [inline client/client.go:171]: dumb question - does a non authenticating client not require apiVersion?

**Comment by hmlanigan** [comment]
In theory, I agree.  In practice, I’m not sure there are OpenStack deployments without username/password or token required?  None of the current goose, nor juju openstack provider test code uses a non authenticating client.  Only the goose client tests do in cases to do with containers, Swift doesn’t use versions.

I need some input about what is supported with openstack deployments used by juju.


On Oct 11, 2016, at 10:03 PM, Ian Booth <notifications@github.com<mailto:notifications@github.com>> wrote:


@wallyworld commented on this pull request.

________________________________

In client/client.go<https://github.com/go-goose/goose/pull/25#pullrequestreview-3798096>:

> @@ -163,44 +170,231 @@ func makeURL(base string, parts []string) string {
        return base + strings.Join(parts, "/")
 }

-func (c *client) MakeServiceURL(serviceType string, parts []string) (string, error) {
+func (c *client) MakeServiceURL(serviceType, apiVersion string, parts []string) (string, error) {
        return makeURL(c.baseURL, parts), nil


dumb question - does a non authenticating client not require apiVersion?

File: client/client.go (hunk 3 above)
Comment (hmlanigan) [reply client/client.go]: No, users can add endpoints with whatever URL they want, with a service type of whatever they want.  One way to pass simplestreams data to Juju is to put it in a Swift container and add an endpoint. Juju explicitly looks for an endpoint with service type "product-streams", and it's tests use other made up names.  We do know that Swift containers are used, so the root url will be the same as the "object-store" url.

File: client/client.go (hunk 4 above)
Comment (hmlanigan) [reply client/client.go:216]: done

File: client/client.go (hunk 5 above)
Comment (hmlanigan) [reply client/client.go]: done

File: client/client.go (hunk 16 above)
Comment (hmlanigan) [reply client/client.go:171]: In theory, I agree.  In practice, I’m not sure there are OpenStack deployments without username/password or token required?  None of the current goose, nor juju openstack provider test code uses a non authenticating client.  Only the goose client tests do in cases to do with containers, Swift doesn’t use versions.

I need some input about what is supported with openstack deployments used by juju.

File: testservices/novaservice/service.go (hunk 14 above)
Comment (hmlanigan) [reply testservices/novaservice/service.go]: I'll add comments to the exported functions I changed the functionality of.

File: client/client.go (hunk 7 above)
Comment (hmlanigan) [reply client/client.go]: done

File: client/client.go (hunk 10 above)
Comment (hmlanigan) [reply client/client.go]: done

File: client/client.go (hunk 11 above)
Comment (hmlanigan) [reply client/client.go]: done

File: testservices/cmd/main.go (hunk 13 above)
Comment (hmlanigan) [reply testservices/cmd/main.go]: removed.

File: client/client.go (hunk 6 above)
Comment (hmlanigan) [reply client/client.go]: per discussion with wallyworld - invalid

File: client/client.go (hunk 9 above)
Comment (hmlanigan) [reply client/client.go]: done

File: client/client.go (hunk 12 above)
Comment (hmlanigan) [reply client/client.go]: done

File: client/client.go (hunk 8 above)
Comment (hmlanigan) [reply client/client.go]: resolved

File: client/client.go (hunk 2 above)
Comment (hmlanigan) [reply client/client.go]: removed, issue resolved with check for containers.

File: testservices/novaservice/service_http.go (hunk 15 above)
Comment (hmlanigan) [reply testservices/novaservice/service_http.go]: done.

File: client/client.go (hunk 6 above)
Comment (wallyworld) [reply client/client.go]: yep, my bad, sorry

File: client/client.go (hunk 16 above)
Comment (wallyworld) [reply client/client.go:171]: Juju itself only uses the authenticating client.

File: client/apiversion.go [hunk 17]
```diff
@@ -0,0 +1,183 @@
...
+		versions = valuesObject.Values
+	} else {
+		if err := json.Unmarshal(Versions, &versions); err != nil {
+			return versions, err
+		}
+	}
+	return versions, nil
+}
+
+// Given a service endpoint, return data on the versions suported by the service
```
Comment (wallyworld) [inline client/apiversion.go]: sorry to be picky - the comment should start with the method name and end with a period. eg
// getAPIVersions returns data on the versions supported by the service at the specified endpoint.

Same comment applies to the getAPIVersionURL method above

File: client/apiversion.go [hunk 18]
```diff
@@ -0,0 +1,183 @@
...
+// and their urls
+func (c *authenticatingClient) getAPIVersions(serviceCatalogURL string) (*apiURLVersion, error) {
+	c.apiVersionMu.Lock()
```
Comment (wallyworld) [inline client/apiversion.go]: This can be moved down to line 154 just before we first start using the apiURLVersions map.
That way we don't prematurely grab the lock

File: client/apiversion.go [hunk 19]
```diff
@@ -0,0 +1,183 @@
...
+	}
+	parsed, err := parseVersion(s)
+	if err != nil {
+		return err
+	}
+	*v = parsed
+	return nil
+}
+
+func parseVersion(s string) (apiVersion, error) {
```
Comment (wallyworld) [inline client/apiversion.go:89]: To aid the reader, a comment here saying that this function parses version strings of the form "v<major>", eg "v2", or "v<major>.<minor>, eg "v2.1"" would be great. Or words to that effect.

**Review by wallyworld — COMMENTED** [review]
Thanks for all the changes. Just add those unit tests as discussed and it should be good.

File: client/apiversion.go (hunk 19 above)
Comment (hmlanigan) [reply client/apiversion.go:89]: updated

File: client/apiversion.go (hunk 18 above)
Comment (hmlanigan) [reply client/apiversion.go]: done.

File: client/apiversion.go (hunk 17 above)
Comment (hmlanigan) [reply client/apiversion.go]: updated.

File: client/local_test.go [hunk 20]
```diff
@@ -97,18 +98,27 @@ func (s *localLiveSuite) SetUpSuite(c *gc.C) {
...
+	if s.authMode != identity.AuthLegacy {
+		s.service.SetupHTTP(nil)
+	}
 	s.LiveTests.SetUpSuite(c)
 }
 
 func (s *localLiveSuite) TearDownSuite(c *gc.C) {
 	s.LiveTests.TearDownSuite(c)
 	s.HTTPSuite.TearDownSuite(c)
+	s.service.Stop()
```
Comment (wallyworld) [inline client/local_test.go]: The service is only started for authMode = AuthLegacy, so stopping it unconditionally doesn't make sense?
It seems that now that there's differences between legacy and non legacy auth, one option is to have a  2 suites, one for legacy auth and one for non-legacy auth. These suites would embed the localLiveSuite suite. The s.service would be moved to legacyLocalLiveSuite and TestAuthenticate would be moved off localLiveSuite and separate implementationd added to each new suite.

File: client/apiversion_test.go [hunk 21]
```diff
@@ -0,0 +1,99 @@
...
+	"net/http"
+	"strconv"
+)
+
+type versionHandler struct {
+	authBody string
+	port     string
+}
+
+type MakeServiceURLTest struct {
```
Comment (wallyworld) [inline client/apiversion_test.go]: does this need to be exported?

File: client/apiversion_test.go [hunk 22]
```diff
@@ -0,0 +1,99 @@
...
+
+const authValuesInformationBody = `{"versions": {"values": [` +
+	`{"status": "stable", "updated": "2015-03-30T00:00:00Z", "media-types": [{"base": "application/json", "type": "application/vnd.openstack.identity-v3+json"}], "id": "v3.4", "links": [{"href": "%s/v3/", "rel": "self"}]},` +
+	`{"status": "stable", "updated": "2014-04-17T00:00:00Z", "media-types": [{"base": "application/json", "type": "application/vnd.openstack.identity-v2.0+json"}], "id": "v2.0", "links": [{"href": "%s/v2.0/", "rel": "self"}, {"href": "http://docs.openstack.org/", "type": "text/html", "rel": "describedby"}]},` +
+	`{"status": "stable", "updated": "2015-03-30T00:00:00Z", "media-types": [{"base": "application/json", "type": "application/vnd.openstack.identity-v3+json"}], "id": "v2.1", "links": [{"href": "%s/v2.1/", "rel": "self"}]}` +
+	`]}}`
+
+func (vh *versionHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("Content-Type", "application/json")
+	//var body []byte
```
Comment (wallyworld) [inline client/apiversion_test.go]: delete me

File: client/apiversion_test.go [hunk 23]
```diff
@@ -0,0 +1,99 @@
...
+	parts       []string
+	success     bool
+	URL         string
+	err         string
+}
+
+func (s *localLiveSuite) makeServiceURLTests() []MakeServiceURLTest {
+	return []MakeServiceURLTest{
+		{
+			serviceType: "compute",
```
Comment (wallyworld) [inline client/apiversion_test.go:28]: Given object-store is special, there should be MakeServiceURL tests for URLs for things other than just compute?

File: client/apiversion_test.go [hunk 24]
```diff
@@ -0,0 +1,99 @@
...
+type MakeServiceURLTest struct {
+	serviceType string
+	version     string
+	parts       []string
+	success     bool
+	URL         string
+	err         string
+}
+
+func (s *localLiveSuite) makeServiceURLTests() []MakeServiceURLTest {
```
Comment (wallyworld) [inline client/apiversion_test.go]: Perhaps this would be best declared near the test cases which make use of it. Maybe move the MakeServiceURL tests to this file?

File: client/apiversion_test.go (hunk 21 above)
Comment (hmlanigan) [reply client/apiversion_test.go]: no - updated

File: client/apiversion_test.go (hunk 24 above)
Comment (hmlanigan) [reply client/apiversion_test.go]: moved TestMakeServiceURL and TestMakeServiceURLValues to apiversion_test.go

File: client/apiversion_test.go (hunk 22 above)
Comment (hmlanigan) [reply client/apiversion_test.go]: done

File: client/apiversion_test.go (hunk 23 above)
Comment (hmlanigan) [reply client/apiversion_test.go:28]: added tests for "object-store" and an "object-store" container but random service type

File: client/apiversion_test.go [hunk 25]
```diff
@@ -0,0 +1,164 @@
...
+			parts:       []string{"foo", "bar"},
+			success:     true,
+			URL:         "http://localhost:%s/swift/v1/foo/bar",
+		},
+		{
+			serviceType: "object-store",
+			version:     "v2.0",
+			parts:       []string{"foo", "bar/"},
+			success:     true,
+			URL:         "http://localhost:%s/swift/v1/foo/bar/",
```
Comment (wallyworld) [inline client/apiversion_test.go:81]: The code comments say that container endpoints have no versioning capability and yet there's a "v1" in the expected URL. Isn't that versioning?  It seems counter intuitive that we can ask for a v2.0 endpoint and get back a v1 URL when that would fail for compute.

Comment (hmlanigan) [reply client/apiversion_test.go:81]: The endpoint says v1, however the API doesn't have a call that returns a list of versions. Here's a link to the API: http://developer.openstack.org/api-ref/object-storage/

For object-storage in MakeServiceURL - the version is ignored.  As far as I'm aware there is not version 2 of the object-store (Swift) API.  It could be that they will create a new endpoint service type if one is needed.  There is already volume and volumev2 service types for Cinder.

You do need to be aware of OpenStack services and version when making calling SendRequest and/or MakeServiceURL.

Comment (wallyworld) [reply client/apiversion_test.go:81]: object-store only has one endpoint. But that endpoint does have a version, "v1".
So we can just have the MakeServiceURL caller in goose (and Juju which calls goose) pass in "" or "v1" and the correct, expected result will be returned. If the caller passes in "v2", that should still error.That allows consistent and understandable behaviour of the API, and keeps the semantics consistent without the special case.

File: client/client.go [hunk 26]
```diff
@@ -163,44 +167,86 @@ func makeURL(base string, parts []string) string {
...
 	}
-	url, ok := c.serviceURLs[serviceType]
+	serviceURL, ok := c.serviceURLs[serviceType]
 	if !ok {
 		return "", errors.New("no endpoints known for service type: " + serviceType)
 	}
-	return makeURL(url, parts), nil
+	requestedVersion, verErr := parseVersion(apiVersion)
+	if serviceType == "object-store" {
+		if apiVersion == "" || (verErr == nil && requestedVersion.major == 1) {
```
Comment (wallyworld) [inline client/client.go]: Oh, I wasn't thinking of this approach. I thought getAPIVersions() would be modified to see if the URL was for object-store and just short circuit without calling the api and return an apiURLVersionInfo with 1.0 URL in it. That to me is the better place to put the short circuit, as it keeps the special case code next to the api in question. The code here is also a lot messier as there are checks for requestedVersion.major == 1 etc which would be avoided if the short circuit were where I am suggesting.
The fact that this works is good, let's not block on that, but my preference would be for the approach I suggest here.

**Review by wallyworld — APPROVED** [review]
I've asked for the short circuit logic to be moved but don't want to block on that as we need to land this. My preference would be to move then logic before landing, no need to wait for another +1

**Comment by hmlanigan** [comment]
The short circuit logic was moved to getAPIVersions() as requested.  Additional tests were added accordingly.

**Comment by hmlanigan** [comment]
$$merge$$

**Comment by axw** [comment]
$$merge$$

**Comment by hmlanigan** [comment]
Updated, for object-store or object-store container endpoints, only “” or “v1” endpoints will be accepted.

On Oct 19, 2016, at 10:17 PM, Ian Booth <notifications@github.com<mailto:notifications@github.com>> wrote:

@wallyworld commented on this pull request.

---

In client/apiversion_test.gohttps://github.com/go-goose/goose/pull/25:

> -                     success:     true,
> -                       URL:         "http://localhost:%s/v2.1/foo/bar/",
> -               },
> -               {
> -                       serviceType: "object-store",
> -                       version:     "",
> -                       parts:       []string{"foo", "bar"},
> -                       success:     true,
> -                       URL:         "http://localhost:%s/swift/v1/foo/bar",
> -               },
> -               {
> -                       serviceType: "object-store",
> -                       version:     "v2.0",
> -                       parts:       []string{"foo", "bar/"},
> -                       success:     true,
> -                       URL:         "http://localhost:%s/swift/v1/foo/bar/",

object-store only has one endpoint. But that endpoint does have a version, "v1".
So we can just have the MakeServiceURL caller in goose (and Juju which calls goose) pass in "" or "v1" and the correct, expected result will be returned. If the caller passes in "v2", that should still error.That allows consistent and understandable behaviour of the API, and keeps the semantics consistent without the special case.
